public/login.js

```diff
@@ -1,26 +1,79 @@
-function login(e) {
-  e.target.action = `https://api.core.ac.uk/login_check?continue=${encodeURIComponent(
-    window.location.origin
-  )}`
-  window.dispatchEvent(new Event('login-in-processing'))
-}
-
 function addLoadingAnimation() {
   const form = document.getElementById('login-form')
   form.classList.add('loading')
 }
 
+function removeLoadingAnimation() {
+  const form = document.getElementById('login-form')
+  form.classList.remove('loading')
+}
+
+function handlePostMessage(event) {
+  if (event.data === 'login-finished') removeLoadingAnimation()
+}
+
+function showWrongCredentialsMessage() {
+  const message = document.getElementById('message')
+  message.innerHTML = 'The username or password you entered is incorrect.'
+  message.classList.add('message-error')
+  message.classList.remove('message-info')
+}
+
+function showLogoutMessage() {
+  const message = document.getElementById('message')
+  message.innerHTML = 'You have been successfully logged out. Log in again.'
+  message.classList.add('message-info')
+  message.classList.remove('message-error')
+}
+
+function login(event) {
+  event.preventDefault()
+  const urlParams = new URLSearchParams(window.location.search)
+  const identityProviderUrl =
+    urlParams.get('identity_provider_url') || 'https://api.core.ac.uk/'
+
+  window.dispatchEvent(new Event('login-processing'))
+  const formData = new FormData(event.target)
+  const data = new URLSearchParams(formData)
+  fetch(`${identityProviderUrl}login_check`, {
+    method: 'POST',
+    body: data,
+    credentials: 'include',
+  })
+    .then(r => {
+      if (r.status === 401) showWrongCredentialsMessage()
+    })
+    .finally(() => {
+      if (window.location !== window.parent.location) {
+        // The page is in an iframe
+        window.parent.postMessage('login-processing', window.location.origin)
```

Comment:
I see no problem if you post message to yourself. It would be just ignored, no need in condition here.

Reply:
Yep, you are right but it doesn't spare any line of code because I decided to support login from /login.html too. So there still will be an if statement.